src/main/java/frc/robot/subsystems/PhoenixDrive.java

```diff
@@ -0,0 +1,77 @@
+import edu.wpi.first.wpilibj.DriverStation.Alliance;
+import edu.wpi.first.wpilibj.Notifier;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.CommandScheduler;
+import edu.wpi.first.wpilibj2.command.Subsystem;
+import java.util.function.Supplier;
+
+public class PhoenixDrive extends SwerveDrivetrain implements Subsystem {
+    private static final double kSimLoopPeriod = 0.005; // 5 ms
+    private Notifier m_simNotifier = null;
+    private double m_lastSimTime;
+
+    /* Blue alliance sees forward as 0 degrees (toward red alliance wall) */
+    private final Rotation2d BlueAlliancePerspectiveRotation = Rotation2d.fromDegrees(0);
+    /* Red alliance sees forward as 180 degrees (toward blue alliance wall) */
+    private final Rotation2d RedAlliancePerspectiveRotation = Rotation2d.fromDegrees(180);
+    /* Keep track if we've ever applied the operator perspective before or not */
+    private boolean hasAppliedOperatorPerspective = false;
+
+    private final SwerveRequest.ApplyChassisSpeeds AutoRequest =
+            new SwerveRequest.ApplyChassisSpeeds();
+
+    private final SwerveRequest.SysIdSwerveTranslation TranslationCharacterization =
+            new SwerveRequest.SysIdSwerveTranslation();
+    private final SwerveRequest.SysIdSwerveRotation RotationCharacterization =
+            new SwerveRequest.SysIdSwerveRotation();
+    private final SwerveRequest.SysIdSwerveSteerGains SteerCharacterization =
+            new SwerveRequest.SysIdSwerveSteerGains();
+
+    public PhoenixDrive(
+            SwerveDrivetrainConstants driveConstants,
+            double odometryUpdateFrequency,
+            SwerveModuleConstants... modules) {
+        super(driveConstants, odometryUpdateFrequency, modules);
+
+        CommandScheduler.getInstance().registerSubsystem(this);
+    }
+
+    public PhoenixDrive(
+            SwerveDrivetrainConstants driveConstants, SwerveModuleConstants... modules) {
+        super(driveConstants, modules);
+
+        CommandScheduler.getInstance().registerSubsystem(this);
+    }
+
+    public Command applyDriveRequest(Supplier<SwerveRequest> requestSupplier) {
+        return run(() -> this.setControl(requestSupplier.get()));
+    }
+
+    public ChassisSpeeds getCurrentSpeeds() {
+        return m_kinematics.toChassisSpeeds(getState().ModuleStates);
+    }
+
+    @Override
+    public void periodic() {
+        if (!hasAppliedOperatorPerspective || DriverStation.isDisabled()) {
+            DriverStation.getAlliance()
+                    .ifPresent(
+                            (color) -> {
+                                this.setOperatorPerspectiveForward(
+                                        color == Alliance.Red
+                                                ? RedAlliancePerspectiveRotation
+                                                : BlueAlliancePerspectiveRotation);
+                            });
+            hasAppliedOperatorPerspective = true;
+        }
```

Comment:
what is this?

Reply:
this was in Phoenix 6's example, I was trying to figure it out myself. I think its for field relative positioning within the SwerveDrivetrain class.

Reply:
okay, I guess we'll see what happens when we test this in sim anyways